File: InjectionIII/FileWatcher.m
//
//  FileWatcher.m
//  InjectionIII
//
//  Created by John Holdsworth on 08/03/2015.
//  Copyright (c) 2015 John Holdsworth. All rights reserved.
//

#import "FileWatcher.h"

@implementation FileWatcher {
    FSEventStreamRef fileEvents;
}

static void fileCallback(ConstFSEventStreamRef streamRef,
                         void *clientCallBackInfo,
                         size_t numEvents, void *eventPaths,
                         const FSEventStreamEventFlags eventFlags[],
                         const FSEventStreamEventId eventIds[]) {
    FileWatcher *self = (__bridge FileWatcher *)clientCallBackInfo;
    // Check that the event flags include an item renamed flag, this helps avoid
    // unnecessary injection, such as triggering injection when switching between
    // files in Xcode.
    BOOL shouldRespondToFileChange = NO;
    for (int i = 0; i < numEvents; i++) {
        uint32 flag = eventFlags[i];
        if (flag & (kFSEventStreamEventFlagItemRenamed | kFSEventStreamEventFlagItemModified)) {
            shouldRespondToFileChange = YES;
            break;
        }
    }

    if (shouldRespondToFileChange == YES) {
        [self performSelectorOnMainThread:@selector(filesChanged:)
                               withObject:(__bridge id)eventPaths waitUntilDone:NO];
    }
}

- (instancetype)initWithRoot:(NSString *)projectRoot plugin:(InjectionCallback)callback;
{
    if ((self = [super init])) {
        self.callback = callback;
        static struct FSEventStreamContext context;
        context.info = (__bridge void *)self;
        fileEvents = FSEventStreamCreate(kCFAllocatorDefault,
                                         fileCallback, &context,
                                         (__bridge CFArrayRef) @[ projectRoot ],
                                         kFSEventStreamEventIdSinceNow, .1,
                                         kFSEventStreamCreateFlagUseCFTypes |
                                         kFSEventStreamCreateFlagFileEvents);
        FSEventStreamScheduleWithRunLoop(fileEvents, CFRunLoopGetMain(), kCFRunLoopDefaultMode);
        FSEventStreamStart(fileEvents);
    }

    return self;
}

- (void)filesChanged:(NSArray *)changes;
{
    NSFileManager *fileManager = [NSFileManager defaultManager];
    NSMutableSet *changed = [NSMutableSet new];

    for (NSString *path in changes) {
        if ([path rangeOfString:INJECTABLE_PATTERN
                        options:NSRegularExpressionSearch].location != NSNotFound &&
            [path rangeOfString:@"DerivedData/|InjectionProject/|main.mm?$"
                        options:NSRegularExpressionSearch].location == NSNotFound &&
            [fileManager fileExistsAtPath:path]) {

            [changed addObject:path];
        }
    }

    //NSLog( @"filesChanged: %@", changed );
    if (changed.count)
        self.callback([[changed objectEnumerator] allObjects]);
}

- (void)dealloc;
{
    FSEventStreamStop(fileEvents);
    FSEventStreamInvalidate(fileEvents);
    FSEventStreamRelease(fileEvents);
}

@end

File: InjectionIII/FileWatcher.swift
//
//  FileWatcher.swift
//  InjectionIII
//
//  Created by John Holdsworth on 08/03/2015.
//  Copyright (c) 2015 John Holdsworth. All rights reserved.
//

import Foundation

let INJECTABLE_PATTERN = "[^~]\\.(mm?|swift|storyboard|xib)$"

public typealias InjectionCallback = (_ filesChanged: NSArray) -> Void

public class FileWatcher: NSObject {
    var fileEvents: FSEventStreamRef! = nil
    var callback: InjectionCallback
    var context = FSEventStreamContext()

    @objc public init(root: String, callback: @escaping InjectionCallback) {
        self.callback = callback
        super.init()
        context.info = unsafeBitCast(self, to: UnsafeMutableRawPointer.self)
        fileEvents = FSEventStreamCreate(kCFAllocatorDefault,
             { (streamRef: FSEventStreamRef,
                clientCallBackInfo: UnsafeMutableRawPointer?,
                numEvents: Int, eventPaths: UnsafeMutableRawPointer,
                eventFlags: UnsafePointer<FSEventStreamEventFlags>,
                eventIds: UnsafePointer<FSEventStreamEventId>) in
                 let watcher = unsafeBitCast(clientCallBackInfo, to: FileWatcher.self)
                 // Check that the event flags include an item renamed flag, this helps avoid
                 // unnecessary injection, such as triggering injection when switching between
                 // files in Xcode.
                 for i in 0 ..< numEvents {
                     let flag = Int(eventFlags[i])
                     if (flag & (kFSEventStreamEventFlagItemRenamed | kFSEventStreamEventFlagItemModified)) != 0 {
                        let changes = unsafeBitCast(eventPaths, to: NSArray.self)
                         DispatchQueue.main.async {
                             watcher.filesChanged(changes: changes)
                         }
                         return
                     }
                 }
             },
             &context, [root] as CFArray,
             FSEventStreamEventId(kFSEventStreamEventIdSinceNow), 0.1,
             FSEventStreamCreateFlags(kFSEventStreamCreateFlagUseCFTypes |
                kFSEventStreamCreateFlagFileEvents))!
        FSEventStreamScheduleWithRunLoop(fileEvents, CFRunLoopGetMain(),
                                         "kCFRunLoopDefaultMode" as CFString)
        FSEventStreamStart(fileEvents)
    }

    func filesChanged(changes: NSArray) {
        var changed = Set<NSString>()

        for path in changes {
            let path = path as! NSString
            if path.range(of: INJECTABLE_PATTERN,
                          options:.regularExpression).location != NSNotFound &&
                path.range(of: "DerivedData/|InjectionProject/|main.mm?$",
                            options:.regularExpression).location == NSNotFound &&
                FileManager.default.fileExists(atPath: path as String) {
                changed.insert(path)
            }
        }

        if changed.count != 0 {
            callback(Array(changed) as NSArray)
        }
    }

    deinit {
        FSEventStreamStop(fileEvents)
        FSEventStreamInvalidate(fileEvents)
        FSEventStreamRelease(fileEvents)
    }
}
